### DISCUSSION TITLE
Копия библиотеки. Нюансы работы

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#195] [Принцип работы]
Для увеличения количества триггеров создал копию библиотеки. Стали появляться ошибки runTasks_. Нашёл здесь где-то в обсуждениях, что отключить runTasks_ можно закомментировав две строки библиотеки. Сделал это в копии библиотеки. Вроде бы ошибки исчезли (пока наблюдаю)... Возник вопрос. Что ещё можно отключить в копии библиотеки, чтобы избежать излишней двойной работы и возможные конфликты всвязи с этим?
Понимаю, что с очередным обновлением библиотеки придётся снова искать определённые строки и каждый раз закомментировать их... Может есть более простые пути?

**Comment by Chimildic** [comment]
Более канонично создавать копию библиотеки при необходимости автоматизации более одного аккаунта Spotify. На такую работу в том числе заточен `Cache` модуль. Например, история прослушиваний сохраняется в разные папки.

Согласно [квоте](https://developers.google.com/apps-script/guides/services/quotas), на одну копию приходится 20 триггеров. Один всегда занят кэшированием истории и лайков. Вряд ли у вас есть задачи, которые загрузят 19 триггеров по 6 минут каждый (ограничение на время выполнения).

Количество используемых триггеров можно оптимизировать: 
- Объединить несколько функций в одну. Сработает при условии, когда плейлисты нужно обновить в один день. Подробно расписано в [экономике триггеров](https://chimildic.github.io/goofy/#/best-practices?id=Экономика-триггеров).
- Дать заданий Клерку. Яркий пример в функции `runTasks_`. Один триггер выполняет три задачи в разное время: каждые 15 минут обновляет историю, каждый день обновляет кэш лайков, каждую неделю перезаписывает весь кэш лайков (на случай удалений из лайков).
- Комбинация экономики и Клерка.
- Перенести редко прослушиваемые плейлисты на ручное обновление через [веб интерфейс](https://github.com/Chimildic/goofy/discussions/9) и/или [триггеры таскера](https://github.com/Chimildic/goofy/discussions/124).

Если все таки делать копию для того же Spotify аккаунта, наверное, нужно закомментировать программное создание триггера для `runTasks_`. Но я не уверен. Какая ошибка происходит?
```js
if (!taskTrigger) {
    // Trigger.create(functionName, 15);
}
```

**Reply by Bondian** [reply]
Закомментил как здесь указано. Пока полёт нормальный, наблюдаю...
После создания копии были ошибки с допуском:
![2022-07-16_06-31-26](https://user-images.githubusercontent.com/85184404/179350860-1bba774a-634f-476e-8cd1-447321efeb26.png)

**Reply by Chimildic** [reply]
Хотите сказать другие триггеры в копии работают? Если у этой функции нет доступа, то и у остальных по идеи не должно быть.

Как копия создавалась? Новое приложение в консоли Spotify делали? Если нет, то получается один и тот же аккаунт пытается авторизоваться в одном и том же приложении. Тогда предыдущий токен должны отозвать.

**Reply by Bondian** [reply]
Делал копию, для того же акrаунта Spotify.
Содавал копию согласно этому совету: "Кроме того, можно создать еще одну копию библиотеки и также получить квоту на 20 триггеров.... Если вам потребовалось создать еще одну копию, можете повторно использовать значения CLIENT_ID и CLIENT_SECRET и не создавать новое приложение на стороне Spotify".  
Насколько я понимаю, если у меня те же ID и SECRET, то новое приложение на стороне Spotify я не создавал? Вроде всё делал по инструкции, пропуская создание нового приложения в spotify dashboard... Но теперь я в этом не уверен... Я в этом не особо соображаю...
Ошибки доступа были не постоянно, раз в сутки, примерно в течении получаса, но появились после создания копии.
Как отозвать токен? Или это не нужно если я использую один Spotify-аккаунт?

**Reply by Chimildic** [reply]
Запустите в редакторе любую функцию с запросом треков в обоих копиях. Ошибка будет?

**Reply by Bondian** [reply]
Нет, ошибкок нет, всё работает

**Reply by Bondian** [reply]
Вот график ошибок после того, как в копии закомментил строку // Trigger.create(functionName, 15);
![2022-07-17_03-09-16](https://user-images.githubusercontent.com/85184404/179373614-7aff5058-c1bc-48f1-a148-5ca24dc40e0c.png)
Видно, что процент ошибок стал стремиться к нулю.
